README.md
# PiGlow

[![Hex.pm Version](https://img.shields.io/hexpm/v/pi_glow.svg?style=flat-square)](https://hex.pm/packages/pi_glow)

PiGlow is a library for controlling a [Pimoroni "PiGlow" LED array](https://shop.pimoroni.com/products/piglow).

With it, you can turn the LEDs on or off, and adjust their power output, allowing your Elixir daemon to give some fun visual feedback to the world around it.

## Installation

PiGlow requires Elixir v1.14.  To use it, add `:pi_glow` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:pi_glow, "~> 0.1.0"}
  ]
end
```

## Usage

```elixir
# Set all LEDs to full power, but turned off:
PiGlow.map_enable_and_power(fn _ -> {false, 255} end)

# Turn on each arm in sequence:
1..3
|> Enum.each(fn arm ->
  PiGlow.map_enable(fn led -> led.arm == arm end)
  Process.sleep(1000)
end)

# Turn everything off:
PiGlow.map_enable_and_power(fn _ -> {false, 0} end)
# Wait before exiting:
PiGlow.wait()
```

More examples can be found in the [examples](https://github.com/wisq/pi_glow/tree/main/examples) directory.

## Documentation

Full documentation can be found at <https://hexdocs.pm/pi_glow>.

## Legal stuff

Copyright © 2023, Adrian Irving-Beer.

PiGlow is released under the [MIT license](https://github.com/wisq/pi_glow/blob/main/LICENSE) and is provided with **no warranty**.  I doubt it's possible to damage your PiGlow with this library, but if you do, I'm not responsible.


CHANGELOG.md
# Changelog

## v0.1.0

* Changelog created.
* Initial release.


config/test.exs
import Config

config :pi_glow,
  start: false,
  start_mock_i2c: true,
  i2c_module: MockI2C

config :logger, level: :info


test/pi_glow_test.exs
defmodule PiGlowDocTest do
  use ExUnit.Case, async: false

  setup :start
  doctest PiGlow

  defp start(_ctx) do
    assert {:ok, _pid} = start_supervised(PiGlow)
    []
  end
end

defmodule PiGlowTest do
  use ExUnit.Case, async: true
  import ExUnit.CaptureLog

  describe "start_link/1" do
    test "opens I2C device and enables I2C writing" do
      assert {:ok, pid} = PiGlow.start_link(name: nil)

      assert i2c = MockI2C.get_device(pid)
      assert i2c.device == "i2c-1"
      assert i2c.writes == [{0x54, <<0, 1>>}]
    end

    test "shuts down if I2C bus is not found" do
      assert {:ignore, log} = with_log(fn -> PiGlow.start_link(name: nil, bus: 99) end)
      assert log =~ ~r{error .* PiGlow .* /dev/i2c-9}x
    end
  end

  describe "stop/0" do
    setup :start

    test "closes I2C", %{pid: pid, i2c: i2c} do
      assert !i2c.closed

      PiGlow.stop(1_000, pid)
      assert i2c = MockI2C.wait_for_close(i2c.ref)
      assert i2c.writes == []
      assert i2c.closed
    end
  end

  describe "set_enable/1" do
    setup :start

    test "sets LED on/off status to given binary values", %{pid: pid} do
      PiGlow.set_enable(<<0x11, 0x22, 0x33>>, pid)
      PiGlow.wait(1_000, pid)

      assert MockI2C.get_device(pid).writes == [
               {0x54, <<0x13, 0x11, 0x22, 0x33>>},
               {0x54, <<0x16, 0xFF>>}
             ]
    end

    test "sets LED on/off status to given values as list", %{pid: pid} do
      values =
        [
          # Byte 1: 0b101010 = 42
          [true, false, true, false, true, false],
          # Byte 2: 0b111111 = 63
          [true, true, true, true, true, true],
          # Byte 3: 0b011001 = 25
          [false, true, true, false, false, true]
        ]
        |> List.flatten()

      PiGlow.set_enable(values, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x13, values_bin::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      assert values_bin == <<42, 63, 25>>
    end
  end

  describe "set_power/1" do
    setup :start

    test "sets LED power to given binary values", %{pid: pid} do
      PiGlow.set_power("eighteenbytebinary", pid)
      PiGlow.wait(1_000, pid)

      assert MockI2C.get_device(pid).writes == [
               {0x54, <<0x01, "eighteenbytebinary">>},
               {0x54, <<0x16, 0xFF>>}
             ]
    end

    test "sets LED power to given values as list", %{pid: pid} do
      values = 0..255 |> Enum.shuffle() |> Enum.take(18)
      PiGlow.set_power(values, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x01, values_bin::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      assert :erlang.binary_to_list(values_bin) == values
    end
  end

  describe "set_enable_and_power/1" do
    setup [:start]

    test "sets LED on/off and power to given binary values", %{pid: pid} do
      PiGlow.set_enable_and_power({<<0x12, 0x23, 0x34>>, "arbitraryPWMvalues"}, pid)
      PiGlow.wait(1_000, pid)

      assert MockI2C.get_device(pid).writes == [
               {0x54, <<0x13, 0x12, 0x23, 0x34>>},
               {0x54, <<0x01, "arbitraryPWMvalues">>},
               {0x54, <<0x16, 0xFF>>}
             ]
    end

    test "sets LED on/off and power to given values as list of tuples", %{pid: pid} do
      values =
        [
          # Byte 1: 0b100110 = 38
          [true: 104, false: 242, false: 173, true: 228, true: 236, false: 164],
          # Byte 2: 0b010011 = 19
          [false: 227, true: 104, false: 8, false: 222, true: 223, true: 29],
          # Byte 3: 0b111101 = 61
          [true: 255, true: 67, true: 119, true: 10, false: 115, true: 252]
        ]
        |> List.flatten()

      PiGlow.set_enable_and_power(values, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x13, enable::binary>>},
               {0x54, <<0x01, power::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      assert enable == <<38, 19, 61>>

      assert power ==
               [
                 <<104, 242, 173, 228, 236, 164>>,
                 <<227, 104, 8, 222, 223, 29>>,
                 <<255, 67, 119, 10, 115, 252>>
               ]
               |> Enum.join("")
    end
  end

  describe "map_enable/1" do
    setup [:start]

    test "calls given function for every LED", %{pid: pid} do
      me = self()

      PiGlow.map_enable(
        fn led ->
          send(me, {:led, led})
          true
        end,
        pid
      )

      PiGlow.LED.leds()
      |> Enum.each(fn led ->
        assert_receive {:led, ^led}
      end)

      refute_receive {:led, _}
    end

    test "sets LED enable based on result of function call", %{pid: pid} do
      PiGlow.map_enable(fn led -> led.colour in [:red, :amber] end, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x13, values::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      # Enabled LED indices: [1, 3, 7, 9, 16, 18]
      assert values == <<0b101000, 0b101000, 0b000101>>
    end
  end

  describe "map_power/1" do
    setup [:start]

    test "calls given function for every LED", %{pid: pid} do
      me = self()

      PiGlow.map_power(
        fn led ->
          send(me, {:led, led})
          0
        end,
        pid
      )

      PiGlow.LED.leds()
      |> Enum.each(fn led ->
        assert_receive {:led, ^led}
      end)

      refute_receive {:led, _}
    end

    test "sets LED power based on result of function call", %{pid: pid} do
      PiGlow.map_power(fn led -> led.arm * 10 + led.ring end, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x01, values::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      assert values == <<36, 35, 34, 33, 12, 13, 16, 15, 14, 11, 21, 22, 31, 23, 32, 24, 25, 26>>
    end
  end

  describe "map_enable_and_power/1" do
    setup [:start]

    test "calls given function for every LED", %{pid: pid} do
      me = self()

      PiGlow.map_enable_and_power(
        fn led ->
          send(me, {:led, led})
          {true, 255}
        end,
        pid
      )

      PiGlow.LED.leds()
      |> Enum.each(fn led ->
        assert_receive {:led, ^led}
      end)

      refute_receive {:led, _}
    end

    test "sets LED power based on result of function call", %{pid: pid} do
      fun = fn led ->
        ena = led.ring in 2..4
        pow = led.ring * 5 + led.arm
        {ena, pow}
      end

      PiGlow.map_enable_and_power(fun, pid)
      PiGlow.wait(1_000, pid)

      assert [
               {0x54, <<0x13, enable::binary>>},
               {0x54, <<0x01, power::binary>>},
               {0x54, <<0x16, 0xFF>>}
             ] = MockI2C.get_device(pid).writes

      # Enabled LED indices: [3, 4, 5, 6, 9, 12, 14, 15, 16]
      assert enable == <<0b001111, 0b001001, 0b011100>>
      assert power == <<33, 28, 23, 18, 11, 16, 31, 26, 21, 6, 7, 12, 8, 17, 13, 22, 27, 32>>
    end
  end

  defp start(_ctx) do
    assert {:ok, pid} = start_supervised({PiGlow, name: nil})
    assert i2c = MockI2C.reset_writes(pid)
    [pid: pid, i2c: i2c, ref: i2c.ref]
  end
end


test/support/mock_i2c.ex
defmodule MockI2C do
  use GenServer

  defmodule DeviceState do
    @enforce_keys [:pid, :ref, :device]
    defstruct(
      pid: nil,
      ref: nil,
      device: nil,
      writes: [],
      closed: false,
      waiting_for_close: nil
    )

    def handle_get(%DeviceState{} = st) do
      %DeviceState{st | writes: Enum.reverse(st.writes)}
    end
  end

  @ets :mock_i2c_pids

  def start_link(_) do
    GenServer.start_link(__MODULE__, nil, name: __MODULE__)
  end

  def open("i2c-99") do
    {:error, :bus_not_found}
  end

  def open(device) do
    GenServer.call(__MODULE__, {:open, self(), device})
  end

  def close(ref) do
    GenServer.cast(__MODULE__, {:close, ref})
  end

  def write(ref, addr, bytes) do
    GenServer.cast(__MODULE__, {:write, ref, addr, bytes})
  end

  def get_device(pid_or_ref) do
    {:ok, i2c} = GenServer.call(__MODULE__, {:get, lookup_ref(pid_or_ref)})
    i2c |> DeviceState.handle_get()
  end

  def reset_writes(pid_or_ref) do
    {:ok, i2c} = GenServer.call(__MODULE__, {:reset_writes, lookup_ref(pid_or_ref)})
    i2c |> DeviceState.handle_get()
  end

  def wait_for_close(pid_or_ref) do
    {:ok, i2c} = GenServer.call(__MODULE__, {:wait_for_close, lookup_ref(pid_or_ref)})
    i2c |> DeviceState.handle_get()
  end

  @impl true
  def init(_) do
    :ets.new(@ets, [:named_table])
    {:ok, %{}}
  end

  @impl true
  def handle_call({:open, pid, device}, _from, states) do
    ref = make_ref()
    dev = %DeviceState{pid: pid, ref: ref, device: device}
    :ets.insert(@ets, {pid, ref})
    {:reply, {:ok, ref}, Map.put(states, ref, dev)}
  end

  @impl true
  def handle_call({:get, ref}, _from, states) do
    {:reply, Map.fetch(states, ref), states}
  end

  @impl true
  def handle_call({:reset_writes, ref}, _from, states) do
    case Map.fetch(states, ref) do
      {:ok, %DeviceState{} = st} ->
        {:reply, {:ok, st}, Map.put(states, ref, %DeviceState{st | writes: []})}

      :error ->
        {:reply, :error, states}
    end
  end

  @impl true
  def handle_call({:wait_for_close, ref}, from, states) do
    case Map.fetch(states, ref) do
      {:ok, %DeviceState{closed: true} = st} ->
        {:reply, {:ok, st}, states}

      {:ok, %DeviceState{closed: false, waiting_for_close: nil} = st} ->
        {:noreply, Map.put(states, ref, %DeviceState{st | waiting_for_close: from})}

      :error ->
        {:reply, :error, states}
    end
  end

  @impl true
  def handle_cast({:close, ref}, states) do
    {:noreply,
     Map.update!(states, ref, fn
       %DeviceState{closed: false, waiting_for_close: waiting} = st ->
         st = %DeviceState{st | closed: true, waiting_for_close: nil}
         if waiting, do: GenServer.reply(waiting, {:ok, st})
         st
     end)}
  end

  @impl true
  def handle_cast({:write, ref, addr, bytes}, states) do
    {:noreply,
     Map.update!(states, ref, fn
       %DeviceState{writes: rest} = st -> %DeviceState{st | writes: [{addr, bytes} | rest]}
     end)}
  end

  defp lookup_ref(pid) when is_pid(pid) do
    [{^pid, ref}] = :ets.lookup(@ets, pid)
    ref
  end

  defp lookup_ref(ref) when is_reference(ref), do: ref
end


mix.exs
defmodule PiGlow.MixProject do
  use Mix.Project

  @github_url "https://github.com/wisq/pi_glow"

  def project do
    [
      app: :pi_glow,
      version: "0.1.0",
      elixir: "~> 1.14",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      description: description(),
      package: package(),
      docs: docs(),
      deps: deps(),
      source_url: @github_url
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {PiGlow.Application, []}
    ]
  end

  defp description do
    """
    PiGlow is a library for controlling a "PiGlow" LED array device.
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README.md", "CHANGELOG.md", "LICENSE"],
      maintainers: ["Adrian Irving-Beer"],
      licenses: ["MIT"],
      links: %{GitHub: @github_url}
    ]
  end

  defp docs do
    [
      main: "readme",
      extras: [
        "README.md",
        "CHANGELOG.md"
      ]
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
      {:circuits_i2c, "~> 1.0"},
      {:ex_doc, "~> 0.27", only: :dev, runtime: false},
      {:mix_test_watch, "~> 1.0", only: :dev, runtime: false},
      {:ex_git_test, "~> 0.1.2", only: [:dev, :test], runtime: false}
    ]
  end
end


examples/demo.exs
#
# Plays a few patterns in sequence.
#
# Run this using `mix run examples/demo.exs`
#

alias PiGlow.LED

long_pulse =
  [0..255, 255..0]
  |> Enum.flat_map(& &1)
  |> Enum.map(&PiGlow.LED.gamma_correct/1)

short_pulse =
  [0..255//3, 255..0//3]
  |> Enum.flat_map(& &1)
  |> Enum.map(&PiGlow.LED.gamma_correct/1)

# Turn on all the LEDs:
PiGlow.map_enable_and_power(fn _ -> {true, 0} end)

# Do three long pulses of all the LEDs at once:
1..3
|> Enum.flat_map(fn _ -> long_pulse end)
|> Enum.each(fn v -> PiGlow.map_power(fn _led -> v end) end)

# Do short pulses by ring -- outside to inside, back to outside -- repeated three times:
[6..1, 2..6, 5..1, 2..6, 5..1, 2..6]
|> Enum.flat_map(&Enum.to_list/1)
|> Enum.each(fn ring ->
  short_pulse
  |> Enum.map(fn v ->
    PiGlow.map_power(fn
      %LED{ring: ^ring} -> v
      _ -> 0
    end)
  end)
end)

# Cycle through the arms, doing short pulses for each in turn, nine times:
1..9
|> Enum.flat_map(fn _ -> 1..3 end)
|> Enum.each(fn arm ->
  short_pulse
  |> Enum.map(fn v ->
    PiGlow.map_power(fn
      %LED{arm: ^arm} -> v
      _ -> 0
    end)
  end)
end)

# All of the above events were delivered asynchronously, as fast as we could
# generate them.
#
# So at this point in the execution, there's a whole bunch of light events
# still in flight to the PiGlow process.  If we just exit here, it will end the
# entire program, and the lights will be in some random state based on where
# they were when we exited.
#
# Let's set all the LEDs to off, and wait for the process to catch up:
PiGlow.map_enable(fn _ -> false end)
IO.puts("All events sent, waiting for completion.")

# Note that we could also run PiGlow.stop() here instead, since we're exiting anyway.
PiGlow.wait()


examples/cycle.exs
#
# Plays a continuous pattern whereby the rings light up in a bouncing in-out-in
# order, leaving a decaying trail behind them.
#
# Run this using `mix run examples/cycle.exs`
#

defmodule Cycler do
  # Decay each ring by 1 brightness per tick.
  @decay_rate 1
  # Move to the next ring every 100 ticks.
  @move_every 50
  # Start at the outermost ring.
  @first_ring 6

  def run do
    :ok = set_led_enable(true)

    spawn_link(fn ->
      1..@move_every
      |> Enum.to_list()
      |> Stream.cycle()
      |> Enum.reduce({@first_ring, [0, 0, 0, 0, 0, 0]}, &run_tick/2)
    end)
  end

  def stop(pid) do
    Process.exit(pid, :normal)
    :ok = set_led_enable(false)
  end

  defp run_tick(1, {ring, ring_values}) do
    # Wait so we don't endlessly fill the message box.
    wait_for_sync()

    ring_values =
      ring_values
      # Decay as normal.
      |> decay_rings()
      # Apply max brightness to the next ring.
      |> List.replace_at(abs(ring) - 1, 255)

    :ok = set_led_power(ring_values)

    {ring |> next_ring(), ring_values}
  end

  defp run_tick(_, {ring, ring_values}) do
    ring_values = ring_values |> decay_rings()

    :ok = set_led_power(ring_values)

    {ring, ring_values}
  end

  defp wait_for_sync do
    {us, :ok} = :timer.tc(PiGlow, :wait, [])
    IO.puts("Synced in #{us}µs  (press enter to exit)")
  end

  # Ring pattern goes in and out, then resets:
  defp next_ring(-5), do: -4
  defp next_ring(-4), do: -3
  defp next_ring(-3), do: -2
  defp next_ring(-2), do: 1
  defp next_ring(1), do: 2
  defp next_ring(2), do: 3
  defp next_ring(3), do: 4
  defp next_ring(4), do: 5
  defp next_ring(5), do: 6
  defp next_ring(6), do: -5

  defp decay_rings(rings) do
    rings
    |> Enum.map(fn
      0 -> 0
      v -> max(0, v - @decay_rate)
    end)
  end

  defp set_led_power(rings) do
    corrected = rings |> Enum.map(&PiGlow.LED.gamma_correct/1)
    PiGlow.map_power(fn led -> Enum.at(corrected, led.ring - 1) end)
  end

  defp set_led_enable(enable) do
    :ok = PiGlow.map_enable(fn _ -> enable end)
    :ok = PiGlow.wait()
  end
end

pid = Cycler.run()
IO.gets("")
Cycler.stop(pid)


lib/pi_glow.ex
defmodule PiGlow do
  require Logger
  @log_prefix "[#{inspect(__MODULE__)}] "

  @leds PiGlow.LED.leds() |> Enum.sort_by(& &1.index)
  @led_count 18 = Enum.count(@leds)
  @enable_chunk_size 6
  @enable_bytes 3 = div(@led_count, @enable_chunk_size)

  @moduledoc """
  A module for changing the power and brightness of the LEDs on a PiGlow device.

  Each PiGlow device contains #{@led_count} LEDs.  (See `PiGlow.LED` for a
  list.)  At any given time, each LED has two hardware properties:

    * **`enable`:** Whether power is being supplied to the LED, expressed as `1`/`true` or `0`/`false`.
    * **`power`:** The amount of power (via [PWM](https://en.wikipedia.org/wiki/Pulse-width_modulation)) being supplied to the LED, expressed as an integer between `0` (no power) and `255` (max power).

  Note that **these properties are independent of each other.**  An LED with
  `enabled` set to `true` may still be turned off if `power` is set to `0`.
  Similarly, an LED with `power` set to `255` will also still be turned off if
  `enable` is set to `false`.

  In general, the easiest way to use this library is to set all LEDs to
  `enable` = `true` when your application starts, and then simply adjust the
  `power` values (using `0` to turn them off).  (This is how most other PiGlow
  libraries work.)  However, having access to both properties allows for useful
  tricks, such as using `set_enable/1` to flash the lights on and off — without
  changing their brightness, and while allowing some lights to remain off
  (`power` = `0`).

  To set these properties, three levels of API are available:

    * **low:** Calling `set_*` with binary arguments.  This will send raw bytes to the device.
    * **medium:** Calling `set_*` with a list of values for each LED (in order).  This will convert those values to the appropriate binaries, and send them using the "low" method above.
    * **high:** Calling `map_*` with a function that maps each `PiGlow.LED` structure to a value.  This will create a list of values, which will be sent via the "medium" method above.

  Thus, you can choose whichever approach works best for your application — for
  example, based on efficiency versus complexity.

  ## Asynchronous API

  With the exception of `start_link/1`, `wait/1` and `stop/1`, all functions in
  this module are asynchronous — they cast messages to the running `PiGlow`
  instance without waiting for a reply, and they always return `:ok`.

  Aside from generally improving performance, this also allows rapidly queuing
  up multiple instructions, e.g. iterating through the full brigtness range
  (from 0 to 255 and back) to "pulse" the LEDs, allowing the PiGlow instance to
  run at full speed without waits.

  In situations where you need to ensure that all your prior instructions have
  completed execution, you can call `wait/1` to send a synchronous request to
  the instance.  This will block until the PiGlow instance has finished
  processing its current message queue.

  ## Process lifecycle

  You generally won't need to start a PiGlow instance manually, as starting
  the application will (by default) automatically start a named instance,
  unless you set `config :pi_glow, start: false` in your application config.

  If you do choose to manually launch a PiGlow instance, all functions in
  this module (besides `start_link`) accept an optional `pid` argument that can
  be used to specify either the PID or registered name of your launched
  instance.

  Note that **launching or shutting down a PiGlow instance does not change the
  state of the LEDs**.  (This is in contrast with other PiGlow libraries, which
  generally apply power to all LEDs on startup, and may also remove power on
  exit.)  Unless you know the LEDs are already enabled for some other reason,
  you'll want to use one of the `*_enable` functions before any `*_power`
  changes will be visible.

  If you are using this library in a script (rather than a daemon), note also
  that **when a script finishes and exits, all in-flight messages are
  discarded**.  If your script makes changes to the LEDs and then immediately
  exits, there's a good chance that most or all of your changes will never get
  processed.  If you want to run LED events just before exiting — for example,
  to turn them off — be sure to use `wait/1` or `stop/1` before exiting.
  """

  use GenServer, restart: :transient
  use PiGlow.AliasI2C

  @default_name __MODULE__

  @default_bus 1
  @bus_addr 0x54
  @cmd_enable_output 0x00
  @cmd_set_pwm_values 0x01
  @cmd_enable_leds 0x13
  @cmd_update 0x16

  @doc """
  Starts a process that will update the PiGlow device LEDs when it receives messages.

  ## Options

    * `:bus` - I2C bus device number (default: `#{@default_bus}`)
    * `:name` - Registered process name to use (default: `#{inspect(@default_name)}`)
      * Use value `nil` to prevent registration.

  This function also accepts all the options accepted by `GenServer.start_link/3`.

  ## Return values

  Same as `GenServer.start_link/3`.
  """
  @spec start_link(GenServer.options()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    {bus_id, opts} = Keyword.pop(opts, :bus, @default_bus)
    opts = Keyword.put_new(opts, :name, @default_name)
    GenServer.start_link(__MODULE__, bus_id, opts)
  end

  @async_return_value """
  Always returns `:ok` immediately (even if there is no `PiGlow` process
  running).  Use `wait/1` if you need to ensure your changes have been sent to
  the device.
  """

  @doc """
  Enables or disables power to each LED.

  The `values` argument can be specified one of two ways:

    * As a list of #{@led_count} booleans, with `true` or `false` indicating
      whether each LED should be enabled or disabled.

    * As a #{@enable_bytes}-byte binary, where each byte is a
      #{@enable_chunk_size}-bit integer indicating which LEDs should be enabled.

  Note that LEDs require both `enabled = true` **and** `power > 0` to light up.
  Use `set_power/1` to adjust LED power, or `set_enabled_and_power/1` to set
  both in a single operation.

  #{@async_return_value}

  ## Examples

      # Enable LEDs at indices 1, 2, 3, 5, 8, 13:
      iex> 1..18 |> Enum.map(&(&1 in [1, 2, 3, 5, 8, 13])) |> PiGlow.set_enable()
      :ok

      # Equivalent to:
      iex> PiGlow.set_enable(<<0b111010, 0b010000, 0b100000>>)
      :ok
  """
  @type enable :: [boolean] | binary
  @spec set_enable(enable, pid) :: :ok
  def set_enable(values, pid \\ @default_name) do
    GenServer.cast(pid, {:set_enable, to_enable_binary(values)})
  end

  defp to_enable_binary(<<binary::binary-size(@enable_bytes)>>), do: binary

  defp to_enable_binary(list) when is_list(list) do
    list
    |> Enum.chunk_every(@enable_chunk_size)
    |> Enum.map(&bools_to_bits/1)
    |> :erlang.list_to_binary()
  end

  @doc """
  Sets the amount of power being delivered to each LED.

  The `values` argument can be specified one of two ways:

    * As a list of #{@led_count} integers, ranging from `0` (off) to `255` (full power).

    * As an #{@led_count}-byte binary, where each byte is an integer, as above.

  Note that LEDs require both `enabled = true` **and** `power > 0` to light up.
  Use `set_enable/1` to turn LEDs on, or `set_enabled_and_power/1` to set both
  in a single operation.

  #{@async_return_value}

  ## Examples

      # Set all LEDs to minimum brightness:
      iex> 1..18 |> Enum.map(fn _ -> 1 end) |> PiGlow.set_power()
      :ok

      # Equivalent to:
      iex> PiGlow.set_power(<<1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1>>)
      :ok
  """
  @type power :: [integer] | binary
  @spec set_power(power, pid) :: :ok
  def set_power(values, pid \\ @default_name) do
    GenServer.cast(pid, {:set_power, to_power_binary(values)})
  end

  defp to_power_binary(<<binary::binary-size(@led_count)>>), do: binary
  defp to_power_binary(list) when is_list(list), do: list |> :erlang.list_to_binary()

  @doc """
  Enables or disables power to all LEDs, and sets the amount of power, in a single operation.

  The `values` argument can be specified one of two ways:

    * As a list of #{@led_count} two-element tuples, each in the format `{enable, power}`, where enable is a boolean and power is an integer.

    * As a two-element tuple, in the format `{enable_values, power_values}`.
      * `enable_values` can be in either format (binary or list) accepted by `set_enable/1`.
      * `power_values` can be in either format (binary or list) accepted by `set_power/1`.

  When using this function, both instructions are sent to the I2C controller,
  one immediately after the other, with no "update" operation sent inbetween.
  This should avoid any unexpected LED flickering caused by setting each value
  independently.

  #{@async_return_value}

  ## Examples

      # Set all LEDs to minimum brightness:
      iex> 1..18 |> Enum.map(fn _ -> 1 end) |> PiGlow.set_power()
      :ok

      # Equivalent to:
      iex> PiGlow.set_power(<<1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1>>)
      :ok
  """
  @spec set_enable_and_power([{boolean, integer}] | {enable, power}, pid) :: :ok
  def set_enable_and_power(values, pid \\ @default_name)

  def set_enable_and_power({enable, power}, pid) do
    GenServer.cast(pid, {:set_enable_and_power, to_enable_binary(enable), to_power_binary(power)})
  end

  def set_enable_and_power([{_, _} | _] = list, pid) do
    Enum.unzip(list)
    |> set_enable_and_power(pid)
  end

  @doc """
  Run a function to determine which LEDs should have power enabled or disabled.

  The `fun` argument must be a function that takes one argument (a `PiGlow.LED`
  structure) and returns a boolean.  The resulting list of booleans will then
  be sent to `set_enable/1`.  See that function for more info.

  Returns `:ok` immediately.

  ## Examples

      # Turn on the green and blue LEDs, turn off the rest:
      iex> PiGlow.map_enable(fn led -> led.colour in [:green, :blue] end)
      :ok

      # Turn on five LEDs at random:
      iex> leds = PiGlow.LED.leds() |> Enum.take_random(5)
      iex> PiGlow.map_enable(&(&1 in leds))
      :ok
  """
  @spec map_enable((PiGlow.LED.t() -> boolean), pid) :: :ok
  def map_enable(fun, pid \\ @default_name) when is_function(fun) do
    @leds
    |> Enum.map(fun)
    |> set_enable(pid)
  end

  @doc """
  Run a function to determine how much power to send to each LED.

  The `fun` argument must be a function that takes one argument (a `PiGlow.LED`
  structure) and returns an integer in the range of `0..255`.  The resulting
  list of integers will then be sent to `set_power/1`.  See that function for
  more info.

  Returns `:ok` immediately.

  ## Examples

      # Set all LEDs to random brightness:
      iex> PiGlow.map_power(fn _ -> Enum.random(1..255) end)
      :ok

      # Set the first arm to max brightness, the second to medium, the third to minimum.
      iex> PiGlow.map_power(fn
      ...>   %{arm: 1} -> 255
      ...>   %{arm: 2} -> 125
      ...>   %{arm: 3} -> 1
      ...> end)
      :ok
  """
  @spec map_power((PiGlow.LED.t() -> integer), pid) :: :ok
  def map_power(fun, pid \\ @default_name) when is_function(fun) do
    @leds
    |> Enum.map(fun)
    |> set_power(pid)
  end

  @doc """
  Run a function to determine the enable status and power to send to each LED.

  The `fun` argument must be a function that takes one argument (a `PiGlow.LED`
  structure) and returns a 2-element tuple `{enable, power}`, where `enable` is
  a boolean and `power` is an integer in the range of `0..255`.

  The resulting list of boolean-integer pairs will be unzipped and sent to
  `set_enable_and_power/1`.  See that function for more info.

  Returns `:ok` immediately.

  ## Examples

      # Enable red LEDs at max brightness, amber at min brightness, disable the rest.
      iex> PiGlow.map_enable_and_power(fn
      ...>   %{colour: :red} -> {true, 255}
      ...>   %{colour: :amber} -> {true, 1}
      ...>   _ -> {false, 0}
      ...> end)
      :ok
  """
  @spec map_enable_and_power((PiGlow.LED.t() -> {boolean, integer}), pid) :: :ok
  def map_enable_and_power(fun, pid \\ @default_name) when is_function(fun) do
    @leds
    |> Enum.map(fun)
    |> set_enable_and_power(pid)
  end

  @doc """
  Waits for all prior messages to be received and processed.

  Returns `:ok` once all pending messages are processed.  (Note that this does
  not guarantee that the PiGlow is idle, only that messages sent prior to the
  start of this `wait` call have been processed.)

  If no message is received within `timeout` milliseconds (default: 60
  seconds), the caller will exit, as per standard `GenServer.call/2` semantics.

  ## Examples

      # Pulse all lights once:
      iex> [0..255, 255..0] |>
      ...>   Enum.flat_map(&Enum.to_list/1) |>
      ...>   Enum.map(&PiGlow.LED.gamma_correct/1) |>
      ...>   Enum.each(fn value ->
      ...>     PiGlow.map_power(fn _ -> value end)
      ...>   end)
      :ok

      # Wait for those 512 events to all be processed:
      iex> PiGlow.wait()
      :ok
  """
  @spec wait(timeout, pid) :: :ok
  def wait(timeout \\ 60_000, pid \\ @default_name) do
    GenServer.call(pid, :wait, timeout)
  end

  @doc """
  Stops a running instance and releases the I2C device.

  Returns `:ok` once all pending messages are processed and the instance has
  been cleanly stopped.

  Note that `PiGlow` uses a default restart policy of `:transient`, meaning
  that it will not be automatically restarted if stopped via this function.

  If no message is received within `timeout` milliseconds (default: 60
  seconds), the caller will exit.

  ## Examples

      # Turn off all LEDs, then shut it down:
      iex> PiGlow.map_enable_and_power(fn _ -> {false, 0} end)
      :ok
      iex> PiGlow.stop()
      :ok
  """
  @spec stop(timeout, pid) :: :ok
  def stop(timeout \\ 60_000, pid \\ @default_name) do
    GenServer.stop(pid, :normal, timeout)
  end

  # --- Internal functions ---

  @impl true
  def init(bus_id) when bus_id in 0..99 do
    case I2C.open("i2c-#{bus_id}") do
      {:ok, bus} ->
        spawn_cleanup(bus)
        :ok = I2C.write(bus, @bus_addr, <<@cmd_enable_output, 1>>)
        {:ok, bus}

      {:error, :bus_not_found} ->
        Logger.error(@log_prefix <> "I2C bus not found: /dev/i2c-#{bus_id}")
        :ignore

      {:error, err} ->
        Logger.error(@log_prefix <> "Unknown error opening I2C bus #{bus_id}: #{inspect(err)}")
        :ignore
    end
  end

  @impl true
  def handle_cast({:set_enable, <<bytes::binary-size(@enable_bytes)>>}, bus) do
    :ok = I2C.write(bus, @bus_addr, <<@cmd_enable_leds>> <> bytes)
    :ok = update(bus)
    {:noreply, bus}
  end

  @impl true
  def handle_cast({:set_power, <<bytes::binary-size(@led_count)>>}, bus) do
    :ok = I2C.write(bus, @bus_addr, <<@cmd_set_pwm_values>> <> bytes)
    :ok = update(bus)
    {:noreply, bus}
  end

  @impl true
  def handle_cast(
        {:set_enable_and_power, <<enable::binary-size(@enable_bytes)>>,
         <<power::binary-size(@led_count)>>},
        bus
      ) do
    :ok = I2C.write(bus, @bus_addr, <<@cmd_enable_leds>> <> enable)
    :ok = I2C.write(bus, @bus_addr, <<@cmd_set_pwm_values>> <> power)
    :ok = update(bus)
    {:noreply, bus}
  end

  @impl true
  def handle_call(:wait, _from, bus) do
    {:reply, :ok, bus}
  end

  defp spawn_cleanup(bus) do
    me = self()

    spawn(fn ->
      ref = Process.monitor(me)

      receive do
        {:DOWN, ^ref, :process, ^me, _} -> I2C.close(bus)
      end
    end)
  end

  defp update(bus), do: I2C.write(bus, @bus_addr, <<@cmd_update, 0xFF>>)

  defp bools_to_bits(list, acc \\ 0)
  defp bools_to_bits([], acc), do: acc
  defp bools_to_bits([true | rest], acc), do: bools_to_bits(rest, 2 * acc + 1)
  defp bools_to_bits([false | rest], acc), do: bools_to_bits(rest, 2 * acc)
end
